src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/Current_Routine_SelectObservations.sql

```diff
@@ -13,24 +13,31 @@ AS
 DECLARE
     @timePeriodCount INT = (SELECT COUNT(year) FROM @timePeriods),
     @paramDefinition NVARCHAR(2000),
-    @sql NVARCHAR(MAX) = N'SELECT o.id ' +
-                         'FROM Observation o ' +
-                         'LEFT JOIN ObservationFilterItem ofi ON o.Id = ofi.ObservationId ' +
-                         'AND ofi.FilterItemId IN (SELECT id FROM @filterItemIds) ' +
-                         'WHERE o.SubjectId = @subjectId ' +
-                         'AND o.LocationId IN (SELECT id FROM @locationIds) '
+    @locationIdsList NVARCHAR(MAX) = (SELECT CONCAT(CONCAT('''', CAST((SELECT STRING_AGG(CAST(Id AS NVARCHAR(MAX)), ''',''') FROM @locationIds) AS NVARCHAR(MAX))), '''')),
+    @sql NVARCHAR(MAX)
+
+    CREATE TABLE #FilterItemId(Id uniqueidentifier)
+    INSERT INTO #FilterItemId SELECT * FROM @filterItemIds
+
+    SET @sql = N'SELECT o.id ' +
+                'FROM Observation o '
+                
+    IF (EXISTS(SELECT * FROM #FilterItemId))
+        SET @sql += N'JOIN ObservationFilterItem ofi ON o.Id = ofi.ObservationId ' +
+                     'JOIN #FilterItemId filterItemId ON ofi.FilterItemId = filterItemId.id '
+
+    SET @sql += N'WHERE o.SubjectId = @subjectId ' +
+                 'AND o.LocationId IN (' + @locationIdsList + ') '
+                
     IF (@timePeriodCount > 0)
         SET @sql += N'AND EXISTS(SELECT 1 from @timePeriods t WHERE t.year = o.Year AND t.timeIdentifier = o.TimeIdentifier) '
-    SET @sql += N'GROUP BY o.Id ' +
-                'HAVING COUNT(DISTINCT ofi.FilterItemId) = (' +
-                '    SELECT COUNT(DISTINCT f.id) AS filterCount' +
-                '    FROM' +
-                '    FilterItem fi' +
-                '    JOIN FilterGroup fg ON fi.FilterGroupId = fg.Id' +
-                '    JOIN Filter f ON fg.FilterId = f.Id' +
-                '    WHERE fi.Id IN (SELECT id FROM @filterItemIds)' +
-                ') ' +
-                'ORDER BY o.Id;'
+
+    IF (EXISTS(SELECT * FROM #FilterItemId))
+        SET @sql += N'GROUP BY o.Id ' +
+                     'HAVING COUNT(DISTINCT ofi.FilterId) = ' +
```

Comment:
Can this lead to incorrect results if the filter items in the query belong to less number of filters than the number of filters of the subject?

If so, because we only enforce one filter item per filter in the UI, I think we should either add a warning in the comments of the stored procedure, or change this back to work as before.

As discussed on Slack, it looks like the right hand side of the `HAVING` clause that we had previously can be worked out earlier in the procedure.

Reply:
Have amended this so that we grab a unique Filter count upfront (the count being based on the unique Filters that appear in the set of FilterItemIds passed in.  Therefore if a Subject has 5 Filters but only the FilterItemIds of 3 of these Filters are passed to the stored proc, the unique Filter count will be 3.

When we do the GROUP BY and the HAVING clauses at the end, we're left with (for each Observation row) all of that row's ObservationFilterItem rows that matched any of the FilterItemIds passed in.  We then just have to get the count of the unique ids of these OFIs' FilterId columns, and check that they match with the unique Filter count.  If there's less, then that means that the Observation row didn't have at least one matching OFI for one of the Filters.